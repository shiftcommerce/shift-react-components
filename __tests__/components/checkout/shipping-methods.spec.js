// Libraries
import React from 'react'

// Components
import ShippingMethods from '../../../src/components/checkout/shipping-methods'

// Objects
import Flash from '../../../src/objects/flash'

// Fixtures
import cartFixture from '../../fixtures/cart'
import shippingMethodsFixture from '../../fixtures/shipping-methods'

describe('ShippingMethods', () => {
  // Setup mock date for testing business days
  const RealDate = Date
  const mockDate = new Date('2019-01-01')

  beforeAll(() => {
    global.Date = class extends Date {
      constructor () {
        super()
        return mockDate
      }
    }
  })

  afterAll(() => {
    global.Date = RealDate
  })

  test('renders a nothing when shipping methods are loading', () => {
    const wrapper = shallow(<ShippingMethods />)

    expect(wrapper.equals(null)).toBe(true)
  })

  test('renders shipping methods when they are passed as props, and their handlers work as expected', () => {
    const mockClassName = 'mock-classname'
    const handleFormSubmit = jest.fn()
    const handleSetShippingMethod = jest.fn()

    const cartShippingMethod = cartFixture.shipping_method
    const cartLineItemsCount = cartFixture.line_items_count

    const wrapper = mount(<ShippingMethods
      cartShippingMethod={cartShippingMethod}
      cartLineItemsCount={cartLineItemsCount}
      className={mockClassName}
      handleFormSubmit={handleFormSubmit}
      handleSetShippingMethod={handleSetShippingMethod}
      shippingMethods={shippingMethodsFixture}
      isThirdPartyPayment={false}
    />)

    // Basic checks for classnames and content
    expect(wrapper.find('div').first()).toHaveClassName(mockClassName)
    expect(wrapper).toIncludeText(shippingMethodsFixture[0].label)
    // Check for estimated delivery
    expect(wrapper).toIncludeText('Wednesday 9th January')

    // Get the last shipping method in the fixture
    const lastShippingMethodFixture = shippingMethodsFixture[shippingMethodsFixture.length - 1]
    // Select the last shipping method
    wrapper.find({ type: 'radio' }).last().prop('onChange')()
    // Check the handler has been called correctly
    expect(handleSetShippingMethod).toHaveBeenCalledWith(lastShippingMethodFixture)

    // Submit the form
    wrapper.find({ role: 'button' }).first().prop('onClick')()
    // Check the handler has been called correctly
    expect(handleFormSubmit).toHaveBeenCalled()
  })
})

test('renders Continue To Payment button when card payments are made', () => {
  const handleFormSubmit = jest.fn()
  const handleSetShippingMethod = jest.fn()
  const cartShippingMethod = cartFixture.shipping_method
  const cartLineItemsCount = cartFixture.line_items_count

  const wrapper = mount(<ShippingMethods
    cartShippingMethod={cartShippingMethod}
    cartLineItemsCount={cartLineItemsCount}
    handleFormSubmit={handleFormSubmit}
    handleSetShippingMethod={handleSetShippingMethod}
    shippingMethods={shippingMethodsFixture}
    isThirdPartyPayment={false}
  />)

  // Basic checks for classnames and content
  expect(wrapper).toIncludeText(shippingMethodsFixture[0].label)
  // Check for estimated delivery
  expect(wrapper.find({ role: 'button' }).first()).toIncludeText('Continue To Payment')
})

test('renders Review Your Order button when third party payment is made', () => {
  const handleFormSubmit = jest.fn()
  const handleSetShippingMethod = jest.fn()
  const cartShippingMethod = cartFixture.shipping_method
  const cartLineItemsCount = cartFixture.line_items_count

  const wrapper = mount(<ShippingMethods
    cartShippingMethod={cartShippingMethod}
    cartLineItemsCount={cartLineItemsCount}
    handleFormSubmit={handleFormSubmit}
    handleSetShippingMethod={handleSetShippingMethod}
    shippingMethods={shippingMethodsFixture}
    isThirdPartyPayment={true}
  />)

  // Basic checks for classnames and content
  expect(wrapper).toIncludeText(shippingMethodsFixture[0].label)
  // Check for estimated delivery
  expect(wrapper.find({ role: 'button' }).first()).toIncludeText('Review Your Order')
})

test('renders error messages', () => {
  // Arrange
  const errorMessage = 'Something went wrong'
  const handleFormSubmit = jest.fn()
  const handleSetShippingMethod = jest.fn()
  const cartShippingMethod = cartFixture.shipping_method
  const cartLineItemsCount = cartFixture.line_items_count

  // Act
  const wrapper = mount(<ShippingMethods
    cartShippingMethod={cartShippingMethod}
    cartLineItemsCount={cartLineItemsCount}
    handleFormSubmit={handleFormSubmit}
    handleSetShippingMethod={handleSetShippingMethod}
    shippingMethods={shippingMethodsFixture}
    isThirdPartyPayment={true}
    errorMessage={errorMessage}
  />)

  // Assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find(Flash).length).toEqual(1)
  expect(wrapper).toIncludeText(errorMessage)
})
