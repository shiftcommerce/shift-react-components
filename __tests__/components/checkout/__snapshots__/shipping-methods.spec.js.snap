// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error messages 1`] = `
<ShippingMethods
  cartLineItemsCount={2}
  cartShippingMethod={
    Object {
      "total": 3.45,
    }
  }
  errorMessage="Something went wrong"
  handleFormSubmit={[MockFunction]}
  handleSetShippingMethod={[MockFunction]}
  isThirdPartyPayment={true}
  shippingMethods={
    Array [
      Object {
        "created_at": "2018-09-24T15:43:27Z",
        "description": "Click & Collect Next Day - shipping method}",
        "id": "18",
        "label": "Click & Collect Next Day",
        "meta_attributes": Object {
          "working_days": Object {
            "data_type": "text",
            "value": "1",
          },
        },
        "reference": "CCND",
        "sku": "EXPRESS_3_DAY",
        "sub_total": 2.08,
        "tax": 0.42,
        "tax_rate": 0.2,
        "total": 2.5,
        "updated_at": "2018-12-14T10:32:31Z",
      },
      Object {
        "created_at": "2018-09-24T15:43:27Z",
        "description": "Click & Collect Standard - shipping method}",
        "id": "19",
        "label": "Click & Collect Standard",
        "meta_attributes": Object {
          "working_days": Object {
            "data_type": "text",
            "value": "2",
          },
        },
        "reference": "CCSD",
        "sku": "EXPRESS_2_DAY",
        "sub_total": 2,
        "tax": 0,
        "tax_rate": 0,
        "total": 2,
        "updated_at": "2018-12-14T10:23:49Z",
      },
      Object {
        "created_at": "2018-09-24T15:43:27Z",
        "description": "Next Day - shipping method}",
        "id": "20",
        "label": "Next Day",
        "meta_attributes": Object {
          "working_days": Object {
            "data_type": "text",
            "value": "1",
          },
        },
        "reference": "NEXT_DAY",
        "sku": "EXPRESS_1_DAY",
        "sub_total": 5,
        "tax": 0,
        "tax_rate": 0,
        "total": 5,
        "updated_at": "2018-12-14T10:24:10Z",
      },
      Object {
        "created_at": "2018-09-24T15:43:27Z",
        "description": "Standard - shipping method}",
        "id": "21",
        "label": "Standard",
        "meta_attributes": Object {
          "working_days": Object {
            "data_type": "text",
            "value": "2",
          },
        },
        "reference": "STANDARD",
        "sku": "REGULAR",
        "sub_total": 1.62,
        "tax": 0.33,
        "tax_rate": 0.2,
        "total": 1.95,
        "updated_at": "2018-12-14T10:32:01Z",
      },
    ]
  }
>
  <div
    aria-label="Shipping Methods"
    className="o-form c-shipping-method"
  >
    <ShippingMethodsHeader
      title="Shipping Method"
    >
      <div
        className="o-form__header c-shipping-method__header"
      >
        <h2>
          Shipping Method
        </h2>
      </div>
    </ShippingMethodsHeader>
    <_default
      modifier="error"
      text="Something went wrong"
    >
      <div
        className="o-flash o-flash--error"
      >
        Something went wrong
      </div>
    </_default>
    <form
      className="o-form__wrapper o-form__background c-shipping-method__list"
    >
      <div
        className="c-shipping-method__list-header"
      >
        Shipping from ... 
        <span
          className="c-shipping-method__list-item-count"
        >
          2 items
        </span>
      </div>
      <div>
        <div
          aria-label="EXPRESS_3_DAY"
          className="o-form__input-group c-shipping-method__list-input-group"
          key="18"
          onClick={[Function]}
        >
          <label
            className="c-shipping-method__radio"
            htmlFor="EXPRESS_3_DAY"
          >
            <input
              checked={false}
              className="c-shipping-method__radio-input"
              id="EXPRESS_3_DAY_18"
              name="shipping_method"
              onChange={[Function]}
              type="radio"
              value="EXPRESS_3_DAY"
            />
            <span
              className="c-shipping-method__radio-caption"
            />
            <span
              className="c-shipping-method__list-cost"
            >
              £
              2.50
            </span>
            <span
              className="c-shipping-method__list-title"
            >
              Click & Collect Next Day
            </span>
            <span
              className="c-shipping-method__list-delivery-date-label"
            >
              Estimated Delivery: 
            </span>
            <span
              className="c-shipping-method__list-delivery-date"
            >
              Monday 13th May
            </span>
          </label>
        </div>
        <div
          aria-label="EXPRESS_2_DAY"
          className="o-form__input-group c-shipping-method__list-input-group"
          key="19"
          onClick={[Function]}
        >
          <label
            className="c-shipping-method__radio"
            htmlFor="EXPRESS_2_DAY"
          >
            <input
              checked={false}
              className="c-shipping-method__radio-input"
              id="EXPRESS_2_DAY_19"
              name="shipping_method"
              onChange={[Function]}
              type="radio"
              value="EXPRESS_2_DAY"
            />
            <span
              className="c-shipping-method__radio-caption"
            />
            <span
              className="c-shipping-method__list-cost"
            >
              £
              2.00
            </span>
            <span
              className="c-shipping-method__list-title"
            >
              Click & Collect Standard
            </span>
            <span
              className="c-shipping-method__list-delivery-date-label"
            >
              Estimated Delivery: 
            </span>
            <span
              className="c-shipping-method__list-delivery-date"
            >
              Tuesday 14th May
            </span>
          </label>
        </div>
        <div
          aria-label="EXPRESS_1_DAY"
          className="o-form__input-group c-shipping-method__list-input-group"
          key="20"
          onClick={[Function]}
        >
          <label
            className="c-shipping-method__radio"
            htmlFor="EXPRESS_1_DAY"
          >
            <input
              checked={false}
              className="c-shipping-method__radio-input"
              id="EXPRESS_1_DAY_20"
              name="shipping_method"
              onChange={[Function]}
              type="radio"
              value="EXPRESS_1_DAY"
            />
            <span
              className="c-shipping-method__radio-caption"
            />
            <span
              className="c-shipping-method__list-cost"
            >
              £
              5.00
            </span>
            <span
              className="c-shipping-method__list-title"
            >
              Next Day
            </span>
            <span
              className="c-shipping-method__list-delivery-date-label"
            >
              Estimated Delivery: 
            </span>
            <span
              className="c-shipping-method__list-delivery-date"
            >
              Monday 13th May
            </span>
          </label>
        </div>
        <div
          aria-label="REGULAR"
          className="o-form__input-group c-shipping-method__list-input-group"
          key="21"
          onClick={[Function]}
        >
          <label
            className="c-shipping-method__radio"
            htmlFor="REGULAR"
          >
            <input
              checked={false}
              className="c-shipping-method__radio-input"
              id="REGULAR_21"
              name="shipping_method"
              onChange={[Function]}
              type="radio"
              value="REGULAR"
            />
            <span
              className="c-shipping-method__radio-caption"
            />
            <span
              className="c-shipping-method__list-cost"
            >
              £
              1.95
            </span>
            <span
              className="c-shipping-method__list-title"
            >
              Standard
            </span>
            <span
              className="c-shipping-method__list-delivery-date-label"
            >
              Estimated Delivery: 
            </span>
            <span
              className="c-shipping-method__list-delivery-date"
            >
              Tuesday 14th May
            </span>
          </label>
        </div>
      </div>
      <div
        className="o-form__input-group"
      >
        <Button
          aria-label="Review Your Order"
          className="o-button--sml"
          label="Review Your Order"
          onClick={[MockFunction]}
          status="positive"
          type="submit"
        >
          <button
            aria-label="Review Your Order"
            className="o-button--sml o-button o-button--positive"
            disabled={false}
            onClick={[MockFunction]}
            role="button"
            tabIndex="0"
            type="submit"
          >
            Review Your Order
          </button>
        </Button>
      </div>
    </form>
  </div>
</ShippingMethods>
`;
